src/libvfcfuncinstrument/libVFCFuncInstrument.cpp

```diff
@@ -0,0 +1,416 @@
+
+#include "../../config.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/CallSite.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include "llvm/Transforms/Utils/UnifyFunctionExitNodes.h"
+#include "llvm/Analysis/TargetLibraryInfo.h"
+#include "llvm/IR/DebugInfoMetadata.h"
+#include "llvm/IR/DebugLoc.h"
+#include "llvm/Analysis/CallGraphSCCPass.h"
+#include "llvm/IR/IntrinsicInst.h"
+#include "llvm/Support/Casting.h" 
+#include "llvm/IR/Metadata.h"
+#include "llvm/IR/CallSite.h"
+#include "llvm/IR/DataLayout.h"
+#include "llvm/IR/Type.h"
+
+#include <set>
+#include <iostream>
+#include <fstream>
+#include <utility>
+#include <stdio.h>
+#include <vector>
+#include <cxxabi.h>
+
+using namespace llvm;
+
+namespace {
+
+  // Enumeration of managed types 
+  enum Ftypes { FLOAT, FLOAT_ARRAY, DOUBLE, DOUBLE_ARRAY};
+
+  struct VfclibFunc : public ModulePass{
+    static char ID;
+
+    VfclibFunc() : ModulePass(ID) {}
+
+    void getAnalysisUsage(AnalysisUsage &AU) const override {
+      AU.setPreservesCFG();
+      AU.addRequired<TargetLibraryInfoWrapperPass>();
+    }
+
+    // Fill have_double and have_float with true if the call_inst pi use at least 
+    // of the managed types
+    void haveFloatingPointArithmetic( Instruction *pi, 
+                                      bool is_from_library, 
+                                      bool is_intrinsic, 
+                                      bool *have_float, 
+                                      bool *have_double,
+                                      Module &M)
+    {
+      Function *f = cast<CallInst>(pi)->getCalledFunction();
+      Type *FloatTy = Type::getFloatTy(M.getContext());
+      Type *DoubleTy = Type::getDoubleTy(M.getContext());
+      Type *FloatPtrTy = Type::getFloatPtrTy(M.getContext());
+      Type *DoublePtrTy = Type::getDoublePtrTy(M.getContext());
+      Type* ReturnTy = pi->getType();
+
+      // Test if return type of pi is float
+      (*have_float) = (ReturnTy == FloatTy || ReturnTy == FloatPtrTy);
+
+      // Test if return type of pi is double
+      (*have_double) = (ReturnTy == DoubleTy || ReturnTy == DoublePtrTy);
+
+      // Test if f treat floats point numbers
+      if (is_intrinsic || is_from_library){
+        // Loop over arguments types 
+        for (auto it = pi->op_begin(); it < pi->op_end()-1; it++){
+          if ((*it)->getType() == FloatTy || (*it)->getType() == FloatPtrTy)
+            (*have_float) = true;
+          if ((*it)->getType() == DoubleTy || (*it)->getType() == DoublePtrTy)
+            (*have_double) = true;
+        }
+      }else{
+        // Loop over each instruction of the function and test if one of them 
+        // use float or double
+        for (auto bbi = f->begin(); bbi != f->end(); bbi++){
```

Comment:
Instead of using iterators, you can use c++11 for-loop ranges:
```
for (auto &s : iteratable) {
```
In many cases it makes code easier to read.